aptos-move/move-examples/token_v2/sources/common.move

```diff
@@ -0,0 +1,513 @@
+            error::permission_denied(ENOT_RECYCLE_ALL_ASSET_OWNER_CAPS)
+        );
+
+        if (has_mint) {
+            let MintCap { asset_addr: _ } = extract(&mut mint);
+        };
+        if (has_freeze) {
+            let FreezeCap { asset_addr: _ } = extract(&mut freeze);
+        };
+        if (has_burn) {
+            let BurnCap { asset_addr: _ } = extract(&mut burn);
+        };
+    }
+
+    /// Borrow a `&Supply` from a fungible asset.
+    public fun borrow_supply<T: key>(asset: &Object<T>): &Supply acquires FungibleAssetMetadata {
+        assert_fungible_asset_metadata_exists(asset);
+        let object_addr = object_address(asset);
+        &borrow_global<FungibleAssetMetadata>(object_addr).supply
+    }
+
+    /// Borrow a `&mut Supply` from a fungible asset.
+    inline fun borrow_supply_mut<T: key>(asset: &Object<T>): &mut Supply acquires FungibleAssetMetadata {
+        assert_fungible_asset_metadata_exists(asset);
+        let object_addr = object_address(asset);
+        &mut borrow_global_mut<FungibleAssetMetadata>(object_addr).supply
+    }
+
+    /// Increase the supply of a fungible asset by minting.
+    public fun increase_supply<T: key>(cap: &MintCap, asset: &Object<T>, amount: u64) acquires FungibleAssetMetadata {
+        assert!(amount > 0, error::invalid_argument(EZERO_AMOUNT));
+        assert_mint_cap_and_asset_match(cap, asset);
+        let supply = borrow_supply_mut(asset);
+        if (option::is_some(&supply.maximum)) {
+            let max = *option::borrow(&supply.maximum);
+            assert!(max - supply.current <= amount, error::invalid_argument(ECURRENT_SUPPLY_OVERFLOW))
+        };
+        supply.current = supply.current + amount;
+    }
+
+    /// Increase the supply of a fungible asset by burning.
+    public fun decrease_supply<T: key>(cap: &BurnCap, asset: &Object<T>, amount: u64) acquires FungibleAssetMetadata {
+        assert!(amount > 0, error::invalid_argument(EZERO_AMOUNT));
+        assert_burn_cap_and_asset_match(cap, asset);
+        let supply = borrow_supply_mut(asset);
+        assert!(supply.current >= amount, error::invalid_argument(ECURRENT_SUPPLY_UNDERFLOW));
+        supply.current = supply.current - amount;
+    }
+
+    /// ================================================================================================================
+    /// Capability functions
+    /// ================================================================================================================
+    struct MintCap has store {
+        asset_addr: address
+    }
+
+    struct FreezeCap has store {
+        asset_addr: address
+    }
+
+    struct BurnCap has store {
+        asset_addr: address
+    }
+
+    struct asset_owner_caps has drop, store {
+        mint: Option<MintCap>,
+        freeze: Option<FreezeCap>,
+        burn: Option<BurnCap>,
+    }
+
+    public fun new_asset_owner_caps(asset_addr: address, cap_flags: vector<bool>): asset_owner_caps {
```

Comment:
As discussed offline, we might be able to remove cap_flags and make this function simpler. Creator/owner can disable mint/burn/freeze later if necessary.